File: 144/binary-tree-preorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function (root) {
    let ans = [];

    function dfs(root, ans) {
        if (!root) return;
        ans.push(root.val);

        dfs(root.left, ans);
        dfs(root.right, ans);
    }

    dfs(root, ans);
    return ans;
};

File: 144/BinaryTreePreorderTraversal.scala
object BinaryTreePreorderTraversal {
  def preorderTraversal(root: TreeNode): List[Int] = {
    @annotation.tailrec
    def preorderTraversal(stack: List[TreeNode], res: List[Int]): List[Int] = stack match {
      case h :: t =>
        val r = h.value :: res
        if (h.right != null && h.left != null)
          preorderTraversal(h.left :: h.right :: t, r)
        else if (h.right != null)
          preorderTraversal(h.right :: t, r)
        else if (h.left != null)
          preorderTraversal(h.left :: t, r)
        else
          preorderTraversal(t, r)

      case _ => res.reverse
    }

    if (root == null) List.empty
    else preorderTraversal(root :: Nil, Nil)
  }

  class TreeNode(var _value: Int = 0) {
    var value: Int = _value
    var left: TreeNode = null
    var right: TreeNode = null
  }
}
